# Exploring Functionality (Folktables)

This notebook shows how each argument changes results and runtime using ACS (Folktables):

- input modes (DataFrame / CSV / two-sample)

- protected_list, continuous_list, fp_map

- n_samples, seed

- method="MSD" vs method="l_inf"

## Imports & basics

In [1]:
import numpy as np
import pandas as pd
from pathlib import Path

from time import perf_counter

from folktables import ACSIncome

from humancompatible.detect import (
    detect_and_score,
    most_biased_subgroup, evaluate_biased_subgroup,
    most_biased_subgroup_csv, evaluate_biased_subgroup_csv,
    most_biased_subgroup_two_samples, evaluate_biased_subgroup_two_samples,
)
from humancompatible.detect.helpers import report_subgroup_bias

import logging
logging.basicConfig(level=logging.INFO, format="[%(levelname)s] %(message)s")

STATE_A = "CA"

In [2]:
# ---------------------------------------------------------------------
# NOTE:
# This notebook is intentionally self-contained. The helper functions
# below are copied from the repository's examples utilities so that the
# notebook can be executed even when downloaded alone (without the full
# repo layout).
#
# Canonical source (kept in sync with the library release):
#   - examples/supports/folktables_utils.py
#   - examples/supports/linf_utils.py
#
# These helpers are for demonstration/visualization only and are not
# part of the public library API.
# ---------------------------------------------------------------------

try:
    from supports.folktables_utils import (
        load_state_data,
        ProtectedOnly,
        FEATURE_PROCESSING,
        CONTINUOUS_FEATURES, 
        FEATURE_NAMES, 
        PROTECTED_VALUES_MAP,
    )
    print("Imported from supports.folktables_utils")
except ModuleNotFoundError:
    print("supports.folktables_utils not found, using local copy")
    from typing import Tuple

    import numpy as np
    import pandas as pd
    from folktables import ACSDataSource, BasicProblem

    """
    You can find all explanation of resources here:
    https://www.icpsr.umich.edu/web/DSDR/studies/25042/datasets/0002/variables/POBP?archive=dsdr

    In "Search Variables" filed, write the code of uknown abbreviation.
    Or you can use:
    https://www2.census.gov/programs-surveys/acs/tech_docs/pums/data_dict/PUMS_Data_Dictionary_2018.pdf
    """

    # Common folktables features that take continuous values
    CONTINUOUS_FEATURES = ["AGEP", "PINCP", "WKHP", "JWMNP", "POVPIP"]

    # Mapping from codes to names
    FEATURE_NAMES = {
        "SEX": "Sex",
        "RAC1P": "Race",
        "AGEP": "Age",
        "MAR": "Marital status",
        "POBP": "Place of birth",
        "DIS": "Disability",
        "CIT": "Citizenship",
        "MIL": "Military service",
        "ANC": "Ancestry",
        "NATIVITY": "Foreign or US native",
        "DEAR": "Difficulty hearing",
        "DEYE": "Difficulty seeing",
        "DREM": "Cognitive difficulty",
        "FER": "Gave birth last year",
        "POVPIP": "Income / Poverty threshold",
        "COW": "Class of worker",
        "SCHL": "Educational attainment",
        "OCCP": "Occupation recode",
        "RELP": "Relationship",
        "WKHP": "Usual hours worked per week past 12 months",
    }

    # Mapping from protected feature values to their interpretation - for visualization
    PROTECTED_VALUES_MAP = {
        "SEX": {1: "Male", 2: "Female"},
        "RAC1P": {
            1: "White",
            2: "Black",
            3: "American Indian",
            4: "Alaska Native",
            5: "Native tribes specified",
            6: "Asian",
            7: "Pacific Islander",
            8: "Some Other Race",
            9: "Two or More Races",
        },
        # "AGEP": {}, Age is not categorical
        "MAR": {
            1: "Married",
            2: "Widowed",
            3: "Divorced",
            4: "Separated",
            5: "Never married",
        },
        # "POBP": {}, Too many values, one for each country
        "_POBP": {
            0: "US",
            1: "Europe",
            2: "Asia",
            3: "Non-US Americas",
            4: "Africa",
            5: "Oceania",
        },
        "DIS": {1: "With a disability", 2: "Without a disability"},
        "CIT": {
            1: "Born in the US",
            2: "Born in external US teritories",
            3: "Born abroad of US parent(s)",
            4: "Naturalized US citizen",
            5: "Not a US citizen",
        },
        "MIL": {
            None: "N/A (<17 years)",
            1: "On active duty",
            2: "No more active duty",
            3: "Active duty for training",
            4: "Never served",
        },
        "ANC": {
            1: "Single",
            2: "Multiple",
            3: "Unclassified",
            4: "Not reported",
            8: "Hidden",
        },
        "NATIVITY": {1: "Native", 2: "Foreign born"},
        "DEAR": {1: "Yes", 2: "No"},
        "DEYE": {1: "Yes", 2: "No"},
        "DREM": {1: "Yes", 2: "No", None: "N/A (<5 years)"},
        "FER": {1: "Yes", 2: "No", None: "N/A"},
        "SCHL": {
            1: "No school completed",
            2: "Nursery school to grade 4",
            3: "Grade 5 or grade 6",
            4: "Grade 7 or grade 8",
            5: "Grade 9",
            6: "Grade 10",
            7: "Grade 11",
            8: "Grade 12 no diploma",
            9: "High school graduate",
            10: "Some college, but less than 1 year",
            11: "One or more years of college, no degree",
            12: "Associate's degree",
            13: "Bachelor's degree",
            14: "Master's degree",
            15: "Professional school degree",
            16: "Doctorate degree",
        },
        "COW": {
            # Not implemented yet. Look at docstrings of this function
        },
        "OCCP": {
            # Not implemented yet. Look at docstrings of this function
        },
        "RELP": {
            # Not implemented yet. Look at docstrings of this function
        },
        "WKHP": {
            # Not implemented yet. Look at docstrings of this function
        },
        # "POVPIP": {}, Poverty ratio has numeric values
    }

    # Some suggested feature simplifications for more statistically useful binning
    FEATURE_PROCESSING = {
        "POBP": lambda x: int(x) // 100,  # group by continents + US
        "OCCP": lambda x: int(x) // 100,
        "PUMA": lambda x: int(x) // 100,
        "POWPUMA": lambda x: int(x) // 1000,
    }

    # A special folktables problem containing all 14 protected attributes used in the MSD paper
    ProtectedOnly = BasicProblem(
        features=[
            "SEX",
            "RAC1P",
            "AGEP",
            "POBP",
            "DIS",
            "CIT",
            "MIL",
            "ANC",
            "NATIVITY",
            "DEAR",
            "DEYE",
            "DREM",
            "FER",
            "POVPIP",
        ],
        target="ST",
        # for fertility and other non-applicable features
        postprocess=lambda x: np.nan_to_num(x, -1),
    )


    def load_state_data(
        state: str,
        problem_cls: BasicProblem = ProtectedOnly,
        survey_year: str = "2018",
        horizon: str = "1-Year",
        survey="person",
        data_root: str = "../data/folktables",
        download: bool = True,
    ) -> Tuple[pd.DataFrame, pd.DataFrame]:
        """
        Download (if needed), load, and project a single-state folktables dataset.

        Args:
            state: Two-letter USPS state code (e.g. "CA", "FL").
            problem_cls: A folktables Problem class (ACSIncome, ACSPublicCoverage, ...).
            survey_year: e.g. "2018".
            horizon: "1-Year" (default) or "5-Year".
            survey: "person" (default) or "household".
            data_root: path under which folktables will write its CSVs.
            download: if True, attempt `get_data(download=True)`, else only local.

        Returns:
            X:      DataFrame with the data.
            y:      DataFrame of the target.
        """
        ds = ACSDataSource(
            survey_year=survey_year,
            horizon=horizon,
            survey=survey,
            root_dir=data_root,
        )
        try:
            raw = ds.get_data(states=[state], download=download)
        except Exception as e:
            print("\n!!!  Automatic download failed:")
            print(f"    {e!r}\n")
            print("→ Please manually download this file and unzip it under:")
            print(f"    {data_root}/{survey_year}/{horizon}/csv_p{state.lower()}.zip")
            print("\nYou can get it from:")
            print(
                f"https://www2.census.gov/programs-surveys/acs/data/pums/{survey_year}/{horizon}/\n"
            )
            raw = ds.get_data(states=[state], download=False)

        X, y, _ = problem_cls.df_to_pandas(raw)
        return X, y


    # ────────── folktables state-code mapping ──────────
    # ACS "place of birth" recodes for states (POBP)
    # https://www.icpsr.umich.edu/web/DSDR/studies/25042/datasets/0002/variables/POBP?archive=dsdr
    STATE_POBP_CODE: dict[str, int] = {
        "AL": 1,  # Alabama
        "AK": 2,  # Alaska
        "AZ": 4,  # Arizona
        "AR": 5,  # Arkansas
        "CA": 6,  # California
        "CO": 8,  # Colorado
        "CT": 9,  # Connecticut
        "DE": 10,  # Delaware
        "DC": 11,  # District of Columbia
        "FL": 12,  # Florida
        "GA": 13,  # Georgia
        "HI": 15,  # Hawaii
        "ID": 16,  # Idaho
        "IL": 17,  # Illinois
        "IN": 18,  # Indiana
        "IA": 19,  # Iowa
        "KS": 20,  # Kansas
        "KY": 21,  # Kentucky
        "LA": 22,  # Louisiana
        "ME": 23,  # Maine
        "MD": 24,  # Maryland
        "MA": 25,  # Massachusetts
        "MI": 26,  # Michigan
        "MN": 27,  # Minnesota
        "MS": 28,  # Mississippi
        "MO": 29,  # Missouri
        "MT": 30,  # Montana
        "NE": 31,  # Nebraska
        "NV": 32,  # Nevada
        "NH": 33,  # New Hampshire
        "NJ": 34,  # New Jersey
        "NM": 35,  # New Mexico
        "NY": 36,  # New York
        "NC": 37,  # North Carolina
        "ND": 38,  # North Dakota
        "OH": 39,  # Ohio
        "OK": 40,  # Oklahoma
        "OR": 41,  # Oregon
        "PA": 42,  # Pennsylvania
        "RI": 44,  # Rhode Island
        "SC": 45,  # South Carolina
        "SD": 46,  # South Dakota
        "TN": 47,  # Tennessee
        "TX": 48,  # Texas
        "UT": 49,  # Utah
        "VT": 50,  # Vermont
        "VA": 51,  # Virginia
        "WA": 53,  # Washington
        "WV": 54,  # West Virginia
        "WI": 55,  # Wisconsin
    }


    def state_to_pobp_code(abbrev: str) -> int:
        """
        Turn a two-letter state code (e.g. 'CA') into the ACS POBP recode.
        Raises a KeyError if the state isn't in the map.
        """
        st = abbrev.strip().upper()
        try:
            return STATE_POBP_CODE[st]
        except KeyError:
            valid = ", ".join(sorted(STATE_POBP_CODE.keys()))
            raise KeyError(
                f"Unknown state abbreviation '{abbrev}'. Valid codes are: {valid}"
            )


Imported from supports.folktables_utils


## Quick API recap

**Core helper**  
`detect_and_score(...) -> (rule, value)` works in 3 modes:
- DataFrame: pass X, y
- CSV: pass csv_path, target_col
- Two-sample: pass X1, X2

Other shared options:
- protected_list, continuous_list, fp_map
- n_samples, seed
- method="MSD" (default) or "l_inf" (+ method_kwargs)

## DataFrame mode (ACSIncome, single state)

In [3]:
# Load ACSIncome for a single state (features & target)
X_df, y_df = load_state_data(STATE_A, problem_cls=ACSIncome)
print(STATE_A, X_df.shape, y_df.shape)
X_df.head()

CA (195665, 10) (195665, 1)


,AGEP,COW,SCHL,MAR,OCCP,POBP,RELP,WKHP,SEX,RAC1P
0,30.0,6.0,14.0,1.0,9610.0,6.0,16.0,40.0,1.0,8.0
1,21.0,4.0,16.0,5.0,1970.0,6.0,17.0,20.0,1.0,1.0
2,65.0,2.0,22.0,5.0,2040.0,6.0,17.0,8.0,1.0,1.0
3,33.0,1.0,14.0,3.0,9610.0,36.0,16.0,40.0,1.0,1.0
4,18.0,2.0,19.0,5.0,1021.0,6.0,17.0,18.0,2.0,1.0


In [4]:
# Start simple: pick a handful of protected features from ACSIncome
candidates = ACSIncome.features
print("ACSIncome feature candidates:", candidates)

protected_list = ["AGEP", "SCHL", "SEX", "RAC1P"]
continuous_list = [c for c in protected_list if c in CONTINUOUS_FEATURES]
fp_map = None
n_samples = 1_000
seed = 42
solver_name = "gurobi"

rule, val = detect_and_score(
    X=X_df, y=y_df,
    protected_list=protected_list,
    continuous_list=continuous_list,
    n_samples=n_samples, seed=seed,
    method="MSD", method_kwargs={"solver": solver_name}
)

report_subgroup_bias(f"{STATE_A} (DataFrame mode)", val, rule, FEATURE_NAMES, PROTECTED_VALUES_MAP)


[INFO] Seeding the run with seed=42 for searching the `rule`.


ACSIncome feature candidates: ['AGEP', 'COW', 'SCHL', 'MAR', 'OCCP', 'POBP', 'RELP', 'WKHP', 'SEX', 'RAC1P']


[INFO] Set parameter Username
[INFO] Set parameter LicenseID to value 2649381
[INFO] Academic license - for non-commercial use only - expires 2026-04-09
[INFO] Seeding the run with seed=42 for searching the `value`.


CA (DataFrame mode)
MSD = 0.243
Rule: AGEP between (np.float64(17.0), np.float64(24.70077))
Explained rule: Age = (np.float64(17.0), np.float64(24.70077))


## Effect of `protected_list`

`protected_list` tells the system **which columns are considered protected attributes** for subgroup search and evaluation. Only these columns are binarized and explored when building a conjunctive rule.

- Adding more protected columns increases the search space (potentially better rules) but also increases runtime.

- Fewer, meaningful protected attributes yield simpler rules that are easier to communicate.

- Default behavior: If `protected_list=None`, all feature columns are treated as protected, which is convenient but may slow the search.

In [5]:
rows = []
for plist in [
    ["AGEP", "SEX"],
    ["AGEP", "SCHL", "SEX", "RAC1P"],
    ["AGEP", "COW", "SCHL", "OCCP", "RELP", "SEX", "RAC1P"],
]:
    print("protected_list =", plist)
    t0 = perf_counter()

    r, v = detect_and_score(
        X=X_df, y=y_df,
        protected_list=plist,
        continuous_list=[c for c in (plist or []) if c in CONTINUOUS_FEATURES],
        n_samples=n_samples, seed=seed,
        method="MSD", method_kwargs={"solver": solver_name}
    )
    rule_str = " AND ".join(str(binop) for _, binop in r)
    
    dt = perf_counter() - t0
    rows.append({
        "protected_list": plist,
        "n_samples": n_samples,
        "MSD": v,
        "rule": rule_str,
        "time_s": dt
    })

pd.DataFrame(rows).sort_values("time_s", ascending=True)


[INFO] Seeding the run with seed=42 for searching the `rule`.


protected_list = ['AGEP', 'SEX']


[INFO] Seeding the run with seed=42 for searching the `value`.
[INFO] Seeding the run with seed=42 for searching the `rule`.


protected_list = ['AGEP', 'SCHL', 'SEX', 'RAC1P']


[INFO] Seeding the run with seed=42 for searching the `value`.
[INFO] Seeding the run with seed=42 for searching the `rule`.


protected_list = ['AGEP', 'COW', 'SCHL', 'OCCP', 'RELP', 'SEX', 'RAC1P']


[INFO] Seeding the run with seed=42 for searching the `value`.


,protected_list,n_samples,MSD,rule,time_s
0,"[AGEP, SEX]",1000,0.242922,"AGEP between (np.float64(17.0), np.float64(24....",0.333644
1,"[AGEP, SCHL, SEX, RAC1P]",1000,0.242922,"AGEP between (np.float64(17.0), np.float64(24....",9.485271
2,"[AGEP, COW, SCHL, OCCP, RELP, SEX, RAC1P]",1000,0.281277,RELP = 0.0,311.397316


## Effect of `continuous_list`

`continuous_list` marks which protected attributes should be treated as **continuous** during binning. Continuous features are automatically partitioned into intervals (e.g., `[a, b)`), so rules can look like `AGEP between (17.0, 24.7)`.

- Intervals can capture targeted ranges (e.g., young adults) that categorical bins would smear out.

- Very fine bins on small samples can be unstable; with larger n_samples results tend to stabilize.

- Default: If a feature is not in `continuous_list`, it's treated as categorical (exact-value equality / inequality).

If you're pre-bucketing a numeric feature via `fp_map` (e.g., decades), leave it out of `continuous_list` so it's treated as categorical.

In [6]:
for cont in [[], ["AGEP"]]:
    print("\ncontinuous_list =", cont)
    r, v = detect_and_score(
        X=X_df, y=y_df,
        protected_list=["AGEP","SCHL","SEX","RAC1P"],
        continuous_list=cont,
        n_samples=n_samples, seed=seed,
        method="MSD", method_kwargs={"solver": solver_name}
    )
    report_subgroup_bias(f"{STATE_A} (continuous_list={cont})", v, r, FEATURE_NAMES, PROTECTED_VALUES_MAP)


[INFO] Seeding the run with seed=42 for searching the `rule`.



continuous_list = []


[INFO] Seeding the run with seed=42 for searching the `value`.
[INFO] Seeding the run with seed=42 for searching the `rule`.


CA (continuous_list=[])
MSD = 0.221
Rule: SCHL = 21.0
Explained rule: Educational attainment = 21.0

continuous_list = ['AGEP']


[INFO] Seeding the run with seed=42 for searching the `value`.


CA (continuous_list=['AGEP'])
MSD = 0.243
Rule: AGEP between (np.float64(17.0), np.float64(24.70077))
Explained rule: Age = (np.float64(17.0), np.float64(24.70077))


## Effect of `fp_map`

`fp_map` is a dict col -> function applied before binarization. Use it to:

- Normalize or clean raw codes (`"Male"/"M"` -> 1, `"Female"/"F"` -> 0).

- Bucket continuous values (e.g., `AGEP` -> decades).

- Collapse rare categories into an `"Other"` bin.

Why it matters:

- Human-meaningful codes produce cleaner rules.

- Well-chosen categories can significantly speed up search without losing signal.

Keep mappings deterministic and simple (no target leakage).

In [7]:
# Bucket ages by decade to simplify
for fp_map_demo in [
    {},
    {
        "AGEP": lambda a: int(a // 10),  # decades
    },
]:
    r, v = detect_and_score(
        X=X_df, y=y_df,
        protected_list=["AGEP","SCHL","SEX","RAC1P"],
        continuous_list=[],     # treat AGEP as categorical after bucketing
        fp_map=fp_map_demo,
        n_samples=n_samples, seed=seed,
        method="MSD", method_kwargs={"solver": solver_name}
    )
    report_subgroup_bias(f"\n{STATE_A} (fp_map={fp_map_demo})", v, r, FEATURE_NAMES, PROTECTED_VALUES_MAP)


[INFO] Seeding the run with seed=42 for searching the `rule`.
[INFO] Seeding the run with seed=42 for searching the `value`.
[INFO] Seeding the run with seed=42 for searching the `rule`.



CA (fp_map={})
MSD = 0.221
Rule: SCHL = 21.0
Explained rule: Educational attainment = 21.0


[INFO] Seeding the run with seed=42 for searching the `value`.



CA (fp_map={'AGEP': <function <lambda> at 0x0000026A024FF9C0>})
MSD = 0.229
Rule: AGEP = 2.0
Explained rule: Age = 2.0


## `n_samples` and `seed`

`n_samples` caps the number of rows via uniform downsampling; `seed` controls that randomness (and any solver randomness we expose).

Why it matters:

- Smaller samples run faster; larger samples give more stable rules/MSD values.

- Fix `seed` to reproduce runs exactly in the notebook.

Rules of thumb:

- Start with a modest `n_samples` (e.g., 1k-5k) to iterate quickly, then increase to validate stability.

- Expect runtime to grow with `n_samples`, and also with the size of `protected_list`.

In [8]:
print("n_samples effect:")
for n in [10, 20, 50, 100, 200, 500, 1_000, 2_000, 5_000, 10_000, 20_000, 50_000]:
    t0 = perf_counter()
    r, v = detect_and_score(
        X=X_df, y=y_df,
        protected_list=["AGEP","SCHL","SEX","RAC1P"],
        continuous_list=["AGEP"],
        n_samples=n, seed=42,
        method="MSD", method_kwargs={"solver": solver_name}
    )
    rule_str = " AND ".join(str(binop) for _, binop in r)
    dt = perf_counter() - t0
    print(f"  n_samples={n:>6} -> MSD={v:.3f}, rule='{rule_str}', time={dt:.3f}s\n")

[INFO] Seeding the run with seed=42 for searching the `rule`.


n_samples effect:


[INFO] Seeding the run with seed=42 for searching the `value`.
[INFO] Seeding the run with seed=42 for searching the `rule`.


  n_samples=    10 -> MSD=0.400, rule='AGEP between (np.float64(17.0), np.float64(24.70077))', time=0.317s



[INFO] Seeding the run with seed=42 for searching the `value`.
[INFO] Seeding the run with seed=42 for searching the `rule`.


  n_samples=    20 -> MSD=0.333, rule='SCHL = 16.0', time=0.424s



[INFO] Seeding the run with seed=42 for searching the `value`.
[INFO] Seeding the run with seed=42 for searching the `rule`.


  n_samples=    50 -> MSD=0.250, rule='SCHL = 16.0', time=0.661s



[INFO] Seeding the run with seed=42 for searching the `value`.
[INFO] Seeding the run with seed=42 for searching the `rule`.


  n_samples=   100 -> MSD=0.242, rule='RAC1P = 1.0 AND SEX = 1.0', time=1.943s



[INFO] Seeding the run with seed=42 for searching the `value`.
[INFO] Seeding the run with seed=42 for searching the `rule`.


  n_samples=   200 -> MSD=0.219, rule='RAC1P = 1.0 AND SEX = 1.0', time=2.301s



[INFO] Seeding the run with seed=42 for searching the `value`.
[INFO] Seeding the run with seed=42 for searching the `rule`.


  n_samples=   500 -> MSD=0.259, rule='SCHL = 21.0', time=3.791s



[INFO] Seeding the run with seed=42 for searching the `value`.
[INFO] Seeding the run with seed=42 for searching the `rule`.


  n_samples=  1000 -> MSD=0.243, rule='AGEP between (np.float64(17.0), np.float64(24.70077))', time=8.988s



[INFO] Seeding the run with seed=42 for searching the `value`.
[INFO] Seeding the run with seed=42 for searching the `rule`.


  n_samples=  2000 -> MSD=0.205, rule='AGEP between (np.float64(17.0), np.float64(24.70077))', time=8.028s



[INFO] Seeding the run with seed=42 for searching the `value`.
[INFO] Seeding the run with seed=42 for searching the `rule`.


  n_samples=  5000 -> MSD=0.202, rule='AGEP between (np.float64(17.0), np.float64(24.70077))', time=26.326s



[INFO] Seeding the run with seed=42 for searching the `value`.
[INFO] Seeding the run with seed=42 for searching the `rule`.


  n_samples= 10000 -> MSD=0.199, rule='AGEP between (np.float64(17.0), np.float64(24.70077))', time=26.097s



[INFO] Seeding the run with seed=42 for searching the `value`.
[INFO] Seeding the run with seed=42 for searching the `rule`.


  n_samples= 20000 -> MSD=0.197, rule='AGEP between (np.float64(17.0), np.float64(24.70077))', time=35.716s



[INFO] Seeding the run with seed=42 for searching the `value`.


  n_samples= 50000 -> MSD=0.196, rule='AGEP between (np.float64(17.0), np.float64(24.70077))', time=90.452s



In [9]:
print("Seed sensitivity:")
for s in [1, 2, 3]:
    t0 = perf_counter()
    r, v = detect_and_score(
        X=X_df, y=y_df,
        protected_list=["AGEP","SCHL","SEX","RAC1P"],
        continuous_list=["AGEP"],
        n_samples=50_000, seed=s,
        method="MSD", method_kwargs={"solver": solver_name}
    )
    rule_str = " AND ".join(str(binop) for _, binop in r)
    dt = perf_counter() - t0
    print(f"  seed={s} -> MSD={v:.3f}, rule={rule_str}, time={dt:.3f}s\n")

[INFO] Seeding the run with seed=1 for searching the `rule`.


Seed sensitivity:


[INFO] Seeding the run with seed=1 for searching the `value`.
[INFO] Seeding the run with seed=2 for searching the `rule`.


  seed=1 -> MSD=0.195, rule=AGEP between (np.float64(17.0), np.float64(24.70077)), time=36.943s



[INFO] Seeding the run with seed=2 for searching the `value`.
[INFO] Seeding the run with seed=3 for searching the `rule`.


  seed=2 -> MSD=0.195, rule=AGEP between (np.float64(17.0), np.float64(24.70077)), time=33.388s



[INFO] Seeding the run with seed=3 for searching the `value`.


  seed=3 -> MSD=0.198, rule=AGEP between (np.float64(17.0), np.float64(24.70077)), time=41.001s



## Method choice: MSD vs ℓ∞

Two complementary questions:

- **MSD (search)**: "_Which subgroup has the largest outcome-rate gap?_"  
Output: a conjunctive rule (e.g., `AGEP between ... AND SEX = Male`) and its discrepancy value (signed or absolute).

- **ℓ∞ (test)**: "_Does this specific subgroup differ from the population by more than δ, in any positive-class feature bin?_"  
Output: `1.0` if the sup-norm distance is ≤ δ (within tolerance), `0.0` otherwise.

When to use:

- Start with MSD to discover a candidate subgroup.

- Use ℓ∞ to validate a particular subgroup against a policy tolerance (e.g., δ = 0.10).

In [10]:
protected_list = ["COW", "SCHL", "OCCP", "SEX", "RAC1P"]
continuous_list = [c for c in protected_list if c in CONTINUOUS_FEATURES]
n_samples = 2_000
seed = 42

rule, val = detect_and_score(
    X=X_df, y=y_df,
    protected_list=protected_list,
    continuous_list=continuous_list,
    n_samples=n_samples, seed=seed,
    method="MSD", method_kwargs={"solver": solver_name}
)
rule_str = " AND ".join(str(binop) for _, binop in rule)

print(rule_str)
print(val)

[INFO] Seeding the run with seed=42 for searching the `rule`.
[INFO] Seeding the run with seed=42 for searching the `value`.


SCHL = 21.0
0.16064135735776874


In [11]:
# MSD is already demonstrated above. Now l_inf:
linf_kwargs = {
    "feature_involved": "SCHL",
    "subgroup_to_check": 21.0,
    "delta": 0.1,
}
_, is_biased = detect_and_score(
    X=X_df, y=y_df,
    protected_list=protected_list,
    continuous_list=continuous_list,
    n_samples=n_samples, seed=seed,
    method="l_inf", method_kwargs=linf_kwargs
)
print("Is the subgroup biased?", is_biased)

[INFO] Seeding the run with seed=42 for searching the `value`.
[INFO] The most impacted subgroup bias <= 0.1


Is the subgroup biased? True


## Data input modes (DataFrame, CSV, two-sample)

- DataFrame mode: You already have `X` and `y` in memory. Fastest for iterative exploration.

- CSV mode: Provide `csv_path` and `target_col` if your data sits on disk. Handy for shareable, reproducible runs.

- Two-sample mode: Provide `X1` and `X2` (same columns). We synthesize a target (`0` for `X1`, `1` for `X2`) to find where the prevalence of a subgroup differs the most across datasets (e.g., state A vs. state B).

In [12]:
n_samples = 1_000
protected_list = ["AGEP", "SCHL", "SEX", "RAC1P"]
continuous_list = [c for c in protected_list if c in CONTINUOUS_FEATURES]

In [13]:
print("=== DataFrame mode ===")
t0 = perf_counter()
rule_df, msd_df = detect_and_score(
    X=X_df, y=y_df,
    protected_list=protected_list,
    continuous_list=continuous_list,
    n_samples=n_samples, seed=seed,
    method="MSD", method_kwargs={"solver": solver_name},
)
dt = perf_counter() - t0
report_subgroup_bias("DataFrame", msd_df, rule_df, FEATURE_NAMES, PROTECTED_VALUES_MAP)
print(f"(elapsed: {dt:.3f}s)\n")

[INFO] Seeding the run with seed=42 for searching the `rule`.


=== DataFrame mode ===


[INFO] Seeding the run with seed=42 for searching the `value`.


DataFrame
MSD = 0.243
Rule: AGEP between (np.float64(17.0), np.float64(24.70077))
Explained rule: Age = (np.float64(17.0), np.float64(24.70077))
(elapsed: 13.146s)



In [14]:
print("=== CSV mode ===")
# Write a small CSV snapshot and run the CSV helpers
csv_path = Path("acs_demo.csv")
pd.concat([X_df, y_df.rename(columns={y_df.columns[0]: "target"})], axis=1).to_csv(csv_path, index=False)

t0 = perf_counter()
rule_csv = most_biased_subgroup_csv(
    csv_path=csv_path, target_col="target",
    protected_list=protected_list,
    continuous_list=continuous_list,
    seed=seed, n_samples=n_samples,
    method="MSD", method_kwargs={"solver": solver_name},
)
msd_csv = evaluate_biased_subgroup_csv(
    csv_path=csv_path, target_col="target",
    protected_list=protected_list,
    continuous_list=continuous_list,
    seed=seed, n_samples=n_samples,
    method="MSD", method_kwargs={"rule": rule_csv},
)
dt = perf_counter() - t0
report_subgroup_bias("CSV", msd_csv, rule_csv, FEATURE_NAMES, PROTECTED_VALUES_MAP)
print(f"(elapsed: {dt:.3f}s)\n")

=== CSV mode ===


[INFO] Seeding the run with seed=42 for searching the `rule`.
[INFO] Seeding the run with seed=42 for searching the `value`.


CSV
MSD = 0.243
Rule: AGEP between (np.float64(17.0), np.float64(24.70077))
Explained rule: Age = (np.float64(17.0), np.float64(24.70077))
(elapsed: 9.646s)



In [15]:
print("=== Two-sample mode ===")
# Compare two states (same columns). Feel free to change STATE_B.
STATE_B = "NY"
X_b, y_b = load_state_data(STATE_B, problem_cls=ACSIncome)

t0 = perf_counter()
# Find the subgroup whose prevalence differs most between X_df (A) and X_b (B)
rule_2s = most_biased_subgroup_two_samples(
    X1=X_df[protected_list], X2=X_b[protected_list],
    protected_list=protected_list,
    continuous_list=continuous_list,
    seed=seed, n_samples=n_samples,
    method="MSD", method_kwargs={"solver": solver_name},
)

# Now quantify that gap on the combined data
msd_2s = evaluate_biased_subgroup_two_samples(
    X1=X_df[protected_list], X2=X_b[protected_list],
    protected_list=protected_list,
    continuous_list=continuous_list,
    seed=seed, n_samples=n_samples,
    method="MSD", method_kwargs={"rule": rule_2s},
)
dt = perf_counter() - t0
report_subgroup_bias(f"Two-sample ({STATE_A} vs {STATE_B})", msd_2s, rule_2s, FEATURE_NAMES, PROTECTED_VALUES_MAP)
print(f"(elapsed: {dt:.3f}s)")

=== Two-sample mode ===


[INFO] Seeding the run with seed=42 for searching the `rule`.
[INFO] Seeding the run with seed=42 for searching the `value`.
[WARNING] Vector y has dtype int64 instead of bool. Assuming value for True is 1.


Two-sample (CA vs NY)
MSD = 0.116
Rule: RAC1P = 1.0
Explained rule: Race = White
(elapsed: 7.943s)


## Solver choice

You can use several MILP solvers through Pyomo. The default **HiGHS** via `appsi_highs` works out-of-the-box. If you have a commercial solver installed, it can speed things up on larger problems.

**Supported names (in our code):**
- `appsi_highs` - default (HiGHS via Appsi); zero-config, solid performance
- `glpk` - fully open source, but usually slowest
- `gurobi` - very fast (requires license)
- `cplex` - commercial, similar to Gurobi (requires license)
- `xpress` - commercial (requires license), or by default "community license"

**Time limits:** we set wall-clock limits only for the five above  
(use `time_limit` to set this limit; defaults to `300` seconds).

Start with `appsi_highs`. If you have a commercial solver available, try it when runtime grows with feature count or sample size.

In [16]:
from pyomo.opt import SolverFactory
candidates = ["appsi_highs", "glpk", "gurobi", "cplex", "xpress"]
{name: SolverFactory(name).available(False) for name in candidates}

c:\personal\work\detect\.venv\Lib\site-packages\pyomo\solvers\plugins\solvers\xpress_direct.py:255: LicenseWarning: Using the Community license in this session. If you have a full Xpress license, pass the full path to your license file to xpress.init(). If you want to use the FICO Community license and no longer want to see this message, use the following code before using the xpress module:
  xpress.init('c:/personal/work/detect/.venv/Lib/site-packages/xpress/license/community-xpauth.xpr')
  xpress.init()


{'appsi_highs': True,
 'glpk': False,
 'gurobi': True,
 'cplex': True,
 'xpress': True}

In [17]:
solver_name = "appsi_highs"  # or "glpk", "gurobi", "cplex", "xpress"
rule, msd = detect_and_score(
    X=X_df, y=y_df,
    protected_list=["AGEP","SCHL","SEX","RAC1P"],
    continuous_list=[c for c in ["AGEP","SCHL","SEX","RAC1P"] if c in CONTINUOUS_FEATURES],
    n_samples=1_000, seed=42,
    method="MSD",
    method_kwargs={"solver": solver_name, "time_limit": 300},
)

[INFO] Seeding the run with seed=42 for searching the `rule`.
[INFO] Running HiGHS 1.11.0 (git hash: 364c83a): Copyright (c) 2025 HiGHS under MIT licence terms
[INFO] RUN!
[INFO] Seeding the run with seed=42 for searching the `value`.


## Verbosity (verbose)

You can control how chatty the library (and, optionally, the MILP solver) is. We use three levels: `0` = silent, `1` = library info/warnings, `2` = library plus solver output (helpful when tuning time_limit/n_min). Defaults to `1`.  
You can pass `verbose` through the parameters to the functions.

In [18]:
for verbose in [0, 1, 2]:
    print(f"\n=== Verbose={verbose} ===")
    rule, msd = detect_and_score(
        X=X_df, y=y_df,
        protected_list=["AGEP","SCHL","SEX","RAC1P"],
        continuous_list=["AGEP"],
        n_samples=2_000, seed=42,
        method="MSD",
        verbose=verbose,
        method_kwargs={"solver": solver_name}
    )


=== Verbose=0 ===


[INFO] Running HiGHS 1.11.0 (git hash: 364c83a): Copyright (c) 2025 HiGHS under MIT licence terms
[INFO] RUN!
[INFO] Seeding the run with seed=42 for searching the `rule`.



=== Verbose=1 ===


[INFO] Running HiGHS 1.11.0 (git hash: 364c83a): Copyright (c) 2025 HiGHS under MIT licence terms
[INFO] RUN!
[INFO] Seeding the run with seed=42 for searching the `value`.
[INFO] Seeding the run with seed=42 for searching the `rule`.



=== Verbose=2 ===


[INFO] Running HiGHS 1.11.0 (git hash: 364c83a): Copyright (c) 2025 HiGHS under MIT licence terms


Running HiGHS 1.11.0 (git hash: 364c83a): Copyright (c) 2025 HiGHS under MIT licence terms


[INFO] ERROR:   setLocalOptionValue: Option "output_flag" cannot be assigned an int
[INFO] MIP  has 32571 rows; 755 cols; 92752 nonzeros; 46 integer variables (46 binary)
[INFO] Coefficient ranges:
[INFO]   Matrix [8e-04, 2e+00]
[INFO]   Cost   [1e+00, 1e+00]
[INFO]   Bound  [1e+00, 1e+00]
[INFO]   RHS    [1e+00, 2e+00]
[INFO] Presolving model
[INFO] 29738 rows, 755 cols, 89212 nonzeros  0s


RUN!
ERROR:   setLocalOptionValue: Option "output_flag" cannot be assigned an int
MIP  has 32571 rows; 755 cols; 92752 nonzeros; 46 integer variables (46 binary)
Coefficient ranges:
  Matrix [8e-04, 2e+00]
  Cost   [1e+00, 1e+00]
  Bound  [1e+00, 1e+00]
  RHS    [1e+00, 2e+00]
Presolving model
29738 rows, 755 cols, 89212 nonzeros  0s


[INFO] 29738 rows, 755 cols, 89212 nonzeros  0s
[INFO] 
[INFO] Solving MIP model with:
[INFO]    29738 rows
[INFO]    755 cols (46 binary, 0 integer, 0 implied int., 709 continuous, 0 domain fixed)
[INFO]    89212 nonzeros


29738 rows, 755 cols, 89212 nonzeros  0s

Solving MIP model with:
   29738 rows
   755 cols (46 binary, 0 integer, 0 implied int., 709 continuous, 0 domain fixed)
   89212 nonzeros


[INFO] 
[INFO] Src: B => Branching; C => Central rounding; F => Feasibility pump; J => Feasibility jump;
[INFO]      H => Heuristic; L => Sub-MIP; P => Empty MIP; R => Randomized rounding; Z => ZI Round;
[INFO]      I => Shifting; S => Solve LP; T => Evaluate node; U => Unbounded; X => User solution;
[INFO]      z => Trivial zero; l => Trivial lower; u => Trivial upper; p => Trivial point
[INFO] 
[INFO]         Nodes      |    B&B Tree     |            Objective Bounds              |  Dynamic Constraints |       Work      
[INFO] Src  Proc. InQueue |  Leaves   Expl. | BestBound       BestSol              Gap |   Cuts   InLp Confl. | LpIters     Time
[INFO] 
[INFO]  J       0       0         0   0.00%   inf             -0                 Large        0      0      0         0     0.8s
[INFO]          0       0         0   0.00%   1               -0                 Large        0      0      0       295     0.9s



Src: B => Branching; C => Central rounding; F => Feasibility pump; J => Feasibility jump;
     H => Heuristic; L => Sub-MIP; P => Empty MIP; R => Randomized rounding; Z => ZI Round;
     I => Shifting; S => Solve LP; T => Evaluate node; U => Unbounded; X => User solution;
     z => Trivial zero; l => Trivial lower; u => Trivial upper; p => Trivial point

        Nodes      |    B&B Tree     |            Objective Bounds              |  Dynamic Constraints |       Work      
Src  Proc. InQueue |  Leaves   Expl. | BestBound       BestSol              Gap |   Cuts   InLp Confl. | LpIters     Time

 J       0       0         0   0.00%   inf             -0                 Large        0      0      0         0     0.8s
         0       0         0   0.00%   1               -0                 Large        0      0      0       295     0.9s


[INFO]  L       0       0         0   0.00%   1               0.2047536167     388.39%    20857     12      0       323    11.3s
[INFO] 
[INFO] Symmetry detection completed in 0.0s
[INFO] Found 1 full orbitope(s) acting on 3 columns
[INFO] 


 L       0       0         0   0.00%   1               0.2047536167     388.39%    20857     12      0       323    11.3s

Symmetry detection completed in 0.0s
Found 1 full orbitope(s) acting on 3 columns



[INFO]          7       0         2  56.25%   1               0.2047536167     388.39%    20851      6     80     39812    19.6s


         7       0         2  56.25%   1               0.2047536167     388.39%    20851      6     80     39812    19.6s


[INFO]         12       0         5  81.25%   1               0.2047536167     388.39%    20856      6    161     44234    26.1s


        12       0         5  81.25%   1               0.2047536167     388.39%    20856      6    161     44234    26.1s


[INFO]         14       0         7 100.00%   0.2047536167    0.2047536167       0.00%    20859      6    164     44429    26.4s
[INFO] 
[INFO] Solving report
[INFO]   Status            Optimal
[INFO]   Primal bound      0.20475361669
[INFO]   Dual bound        0.20475361669
[INFO]   Gap               0% (tolerance: 0.01%)
[INFO]   P-D integral      58.675079759
[INFO]   Solution status   feasible
[INFO]                     0.20475361669 (objective)
[INFO]                     0 (bound viol.)
[INFO]                     0 (int. viol.)
[INFO]                     0 (row viol.)
[INFO]   Timing            26.36 (total)
[INFO]                     0.00 (presolve)
[INFO]                     0.00 (solve)
[INFO]                     0.00 (postsolve)
[INFO]   Max sub-MIP depth 2
[INFO]   Nodes             14
[INFO]   Repair LPs        0 (0 feasible; 0 iterations)
[INFO]   LP iterations     44429 (total)
[INFO]                     3545 (strong br.)
[INFO]                     28 (separation)
[INFO]  

        14       0         7 100.00%   0.2047536167    0.2047536167       0.00%    20859      6    164     44429    26.4s

Solving report
  Status            Optimal
  Primal bound      0.20475361669
  Dual bound        0.20475361669
  Gap               0% (tolerance: 0.01%)
  P-D integral      58.675079759
  Solution status   feasible
                    0.20475361669 (objective)
                    0 (bound viol.)
                    0 (int. viol.)
                    0 (row viol.)
  Timing            26.36 (total)
                    0.00 (presolve)
                    0.00 (solve)
                    0.00 (postsolve)
  Max sub-MIP depth 2
  Nodes             14
  Repair LPs        0 (0 feasible; 0 iterations)
  LP iterations     44429 (total)
                    3545 (strong br.)
                    28 (separation)
                    33829 (heuristics)
